### PR TITLE
Update cluster state with type mapping also for failed indexing request

### LINKED ISSUE #8650: mapping potentially lost with `"dynamic" : "strict"`, `_default_` mapping and failed document index
The effect is very similar to https://github.com/elasticsearch/elasticsearch/pull/5623
When a document is indexed that does have a dynamic field then the indexing fails as expected. 
However, the type is created locally in the mapper service of the node but never updated on master, see https://github.com/brwe/elasticsearch/commit/340f5c5de207a802085f23aeb984dbd98349301a#diff-defbaaff93b959a2f9a93e7167f6f345R165

This can cause several problems:

1. `_default_` mappings are applied locally and can potentially later not be updated anymore, see https://github.com/brwe/elasticsearch/commit/340f5c5de207a802085f23aeb984dbd98349301a#diff-ed65252ffbbf8656bf257a8cd6251420R68 (thanks @pkoenig10 for the test, https://github.com/elasticsearch/elasticsearch/issues/8423#issuecomment-64395503)
2. Mappings that were created via `_default_` mappings when indexing a document can be lost, see https://github.com/brwe/elasticsearch/commit/340f5c5de207a802085f23aeb984dbd98349301a#diff-defbaaff93b959a2f9a93e7167f6f345R187

### PR DESCRIPTION
**Pull request by brwe**
When indexing of a document with a type that is not in the mappings fails,
for example because "dynamic": "strict" but doc contains a new field,
then the type is still created on the node that executed the indexing request.
However, the change was never added to the cluster state.
This commit makes sure mapping updates are always added to the cluster state
even if indexing of a document fails.

closes #8650

### CODE REVIEW

**Comment by brwe** [comment]
Updated. I also moved the integration test to a separate class as DedicatedAggregationTests is hardly the right place.

File: src/main/java/org/elasticsearch/action/WriteFailure.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+package org.elasticsearch.action;
+
+import org.elasticsearch.ElasticsearchException;
+import org.elasticsearch.ElasticsearchWrapperException;
+import org.elasticsearch.common.Nullable;
+
+
+public class WriteFailure extends ElasticsearchException implements ElasticsearchWrapperException {
+    @Nullable
```
Comment (rjernst) [inline src/main/java/org/elasticsearch/action/WriteFailure.java]: Why is this nullable when there is an assert forbidding null in the ctor?

File: src/main/java/org/elasticsearch/action/WriteFailure.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+ */
+
+package org.elasticsearch.action;
+
+import org.elasticsearch.ElasticsearchException;
+import org.elasticsearch.ElasticsearchWrapperException;
+import org.elasticsearch.common.Nullable;
+
+
+public class WriteFailure extends ElasticsearchException implements ElasticsearchWrapperException {
```
Comment (rjernst) [inline src/main/java/org/elasticsearch/action/WriteFailure.java]: Do we have any other exception classes that don't end in "Exception"? We should probably stick with that naming scheme..

File: src/main/java/org/elasticsearch/action/WriteFailure.java (hunk 1 above)
Comment (rjernst) [inline src/main/java/org/elasticsearch/action/WriteFailure.java]: ignore, im blind. :)

File: src/main/java/org/elasticsearch/action/WriteFailure.java (hunk 2 above)
Comment (rjernst) [inline src/main/java/org/elasticsearch/action/WriteFailure.java]: And I see that this was just extracted out from an inner class, but I think making it public means we should be better with the name.

File: src/main/java/org/elasticsearch/action/index/TransportIndexAction.java [hunk 3]
```diff
@@ -185,39 +187,52 @@ protected ShardIterator shards(ClusterState clusterState, InternalRequest reques
...
-            indexShard.index(index);
-            version = index.version();
-            created = index.created();
-        } else {
-            Engine.Create create = indexShard.prepareCreate(sourceToParse,
-                    request.version(), request.versionType(), Engine.Operation.Origin.PRIMARY, request.canHaveDuplicates(), request.autoGeneratedId());
-            if (create.parsedDoc().mappingsModified()) {
-                mappingUpdatedAction.updateMappingOnMaster(shardRequest.shardId.getIndex(), create.docMapper(), indexService.indexUUID());
+
+        Engine.IndexingOperation op;
```
Comment (rjernst) [inline src/main/java/org/elasticsearch/action/index/TransportIndexAction.java]: I'm not seeing where this is used after it is set?

**Comment by rjernst** [comment]
@brwe I started looking at this again, but then realized I'm not sure this is what we want.  Isn't the point of dynamic: strict to *not* modify mappings?  Adding an empty type in this case seems counter intuitive...

**Comment by brwe** [comment]
> Isn't the point of dynamic: strict to not modify mappings?

`dynamic: strict` means that no fields may be added that were not defined before. The setting for strictly not adding pre defined types is `"index.mapper.dynamic": false`. 

However, I agree it is weird that the mapping is created although the document is not indexed.  We create mappings in some places where we then do not index, like index failures in bulk requests (see https://github.com/elasticsearch/elasticsearch/pull/5623) or when percolating a document that has a type which was not defined before.
The reason why we do this is that currently mappings are added to the node local DocumentMapper the moment we parse. This means that we have to update the mapping, regardless of if the document is indexed or not because else the mapping in cluster state and in DocumentMapper are out of sync and therefore the mapping is lost the moment we restart the node. 
While reading the code I actually found another example where this causes problems: When we parse a document halfway and then get a parsing exception, the parsed fields that were added before are then also not added to the cluster state and can be lost when restarting the node. I added a test and fix to this pr.
Now, I think we have two options:
- fix: Parsing of documents that are not indexed should not change the Document mapper or the mapping in the cluster state 
- push the fixes as is which would at least prevent the mappings to be lost and live with the fact that mappings are created even if indexing of the doc fails.

I would like to do the latter and work on the first in a different issue and pr because I could imagine this needs a little more work.

**Comment by rjernst** [comment]
I'm all for making the behavior consistent and predictable.  What concerns me is the future, when mappings are cleaned up and immutable (#9365) so that we can fix all of these issues you mention and not modify mappings when failures occur.  I do not want someone to then argue the current behavior needs to be maintained. So can we downplay this in the docs somehow? I don't want anyone to rely on this behavior, or even have the expectation of relying on the behavior.

**Comment by rjernst** [comment]
Obviously this PR has no doc changes.  I guess what I mean is, can we maintain the fact that this behavior cannot be relied upon if someone comes along and says "what is the behavior and why isn't it documented?".

This question aside, the change LGTM.

**Comment by brwe** [comment]
ok, changed WriteFailure -> WriteFailureException. Will push in a minute to 1.4 1.x and master

**Comment by brwe** [comment]
spoke too soon. the second fix should also be in prepareIndex and not just prepareCreate. Also, I am not sure if it is sufficient to check if it was a MapperParsingException  but also we should probably check if the mapping was actually modified before we update on master.

**Comment by brwe** [comment]
I'll push only the original fix (mapping lost with dynamic_strict...) and make a separate issue for the second one.

**Comment by javanna** [comment]
@brwe can you please label? :)

**Pull request closed without merging**